persistence/pre_persistence/persistence.go

```diff
@@ -0,0 +1,306 @@
+		// chunk into 100000 byte segments
+		if len(result) >= 100000 {
+			result = crypto.SHA3Hash(result)
+		}
+	}
+	it.Release()
+	// potential for double hash here
+	return crypto.SHA3Hash(result), nil
+}
+
+// Reset to the first save point
+func (m *PrePersistenceContext) Reset() error {
+	return m.RollbackToSavePoint(FirstSavePointKey)
+}
+
+// Commit the KV pairs to the parent (commit) db
+func (m *PrePersistenceContext) Commit() error {
+	index := len(m.DBs) - 1
+	db := m.DBs[index]
+	it := db.NewIterator(&util.Range{})
+	it.First()
+	for ; it.Valid(); it.Next() {
+		if err := m.Parent.GetCommitDB().Put(HeightKey(m.Height, it.Key()), it.Value()); err != nil {
+			return err
+		}
+	}
+	it.Release()
+	m.Release()
+	parentIt := m.Parent.GetCommitDB().NewIterator(&util.Range{
+		Start: HeightKey(m.Height, nil),
+		Limit: PrefixEndBytes(HeightKey(m.Height, nil)),
+	})
+	parentIt.First()
+	m.Height = m.Height + 1
+	// copy over the entire last height
+	for ; parentIt.Valid(); parentIt.Next() {
+		newKey := HeightKey(m.Height, KeyFromHeightKey(parentIt.Key()))
+		if err := m.Parent.GetCommitDB().Put(newKey, parentIt.Value()); err != nil {
+			return err
+		}
+	}
+	parentIt.Release()
+	return nil
+}
+
+func (m *PrePersistenceContext) Release() {
+	m.SavePoints = nil
+	for _, db := range m.DBs {
+		db.Reset()
+	}
+	m.DBs = nil
+	return
+}
+
+// Store returns the latest 'app state' db object
+func (m *PrePersistenceContext) Store() *memdb.DB {
+	i := len(m.DBs) - 1
+	if i < 0 {
+		panic(fmt.Errorf("zero length mock persistence context"))
+	}
+	return m.DBs[i]
+}
+
+func (m *PrePersistenceContext) GetHeight() (int64, error) {
+	return m.Height, nil
+}
+
+func (m *PrePersistenceContext) GetBlockHash(height int64) ([]byte, error) {
+	db := m.Store()
+	block := Block{}
+	key := append(BlockPrefix, []byte(fmt.Sprintf("%d", height))...)
+	val, err := db.Get(key)
+	if err != nil {
+		return nil, err
+	}
+	if err := proto.Unmarshal(val, &block); err != nil {
+		return nil, err
+	}
+	return []byte(block.BlockHeader.Hash), nil
+}
+
+func (m *PrePersistenceContext) TransactionExists(transactionHash string) bool {
+	db := m.Store()
+	return db.Contains(append(TransactionKeyPrefix, []byte(transactionHash)...))
+}
+
+func NewMemDB() *memdb.DB {
+	return memdb.New(comparer.DefaultComparer, 100000)
+}
+
+func CopyMemDB(src, dest *memdb.DB) error {
+	it := src.NewIterator(&util.Range{})
+	it.First()
+	defer it.Release()
+	for ; it.Valid(); it.Next() {
+		err := dest.Put(it.Key(), it.Value())
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+var (
+	elenEncoder = lexnum.NewEncoder('=', '-')
+)
+
+func HeightKey(height int64, k []byte) (key []byte) {
+	keyString := fmt.Sprintf("%s/%s", elenEncoder.EncodeInt(int(height)), k)
+	return []byte(keyString)
+}
+
+func KeyFromHeightKey(heightKey []byte) (key []byte) {
+	k := strings.SplitN(string(heightKey), "/", 2)[1]
+	return []byte(k)
+}
+
+func PrefixEndBytes(prefix []byte) []byte {
```

Comment:
Since this function has no tests, I think a simple comment showing an example input & output, or an explanation would make it easier to understand.